### ISSUE TITLE
Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused

### ISSUE THREAD

**Issue by Phandora** [wazuh/wazuh#7968]
|Wazuh version|Component|
|---|---|
| v4.0.4-40011 | - |
| v4.1.2-40111 | - |

Hi,

When using the `<disabled>yes</disabled>` setting in the **active-response** block,  modifying the `ossec.conf` file from the Kibana interface will generate the following error:

```
{"statusCode":500,"error":"Internal Server Error","message":"3013 - Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused"}
```

It also happens when modifying `custom rules`:
```
File local_rules.xml saved, but there were found several error while validating the configuration.

3013 - Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused
```

In the **api logs** we can see the following error:

```bash
2021/03/22 10:03:40 ERROR: Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/manager.py", line 300, in validate_ossec_conf
    execq_socket.connect(execq_socket_path)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/dapi/dapi.py", line 234, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
  File "/var/ossec/framework/python/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/dapi/dapi.py", line 208, in run_local
    data = self.f(**self.f_kwargs)
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/rbac/decorators.py", line 417, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/manager.py", line 306, in validation
    response = validate_ossec_conf()
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/manager.py", line 303, in validate_ossec_conf
    raise WazuhInternalError(1013, extra_message=extra_msg)
wazuh.core.exception.WazuhInternalError: Error 1013 - Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused

```

> If the disabled setting in the **active-response** block is changed from `yes` to `no`, the error is not shown.

<details>
<summary>Screenshots</summary>

![image](https://user-images.githubusercontent.com/33484080/111987379-78a85480-8b0f-11eb-9917-b0c2f0249b0e.png)

![image](https://user-images.githubusercontent.com/33484080/111987440-88c03400-8b0f-11eb-92f0-0b06b914a6dc.png)

</details>

<details>
<summary>Active-response blocks</summary>

```
  <active-response>
    <disabled>yes</disabled>
    <command>firewall-drop</command>
    <location>local</location>
  </active-response>
  <active-response>
    <disabled>yes</disabled>
    <command>firewall-drop</command>
    <location>local</location>
    <timeout>6000000</timeout>
  </active-response>
  <active-response>
    <disabled>no</disabled>
    <command>firewall-drop</command>
    <rules_id>31103,31151,31152,31153,31154,31164,5712,100100,100101,100102</rules_id>
    <location>local</location>
    <timeout>6000000</timeout>
  </active-response>
```

</details>

**Comment by vicferpoy** [comment]
## Issue update

Hello team,

After investigating this issue, I came to the conclusion that this does not have anything to do with file endpoints. This is strictly related to endpoints that use the `WAZUH_PATH/queue/alerts/execq` socket.

The `GET /manager/configuration` and `PUT /rules/files` endpoints fail because in Kibana there is a double check and the `GET /manager/configuration/validation` endpoints is always called after these ones. 

As we can see here, the `GET /manager/configuration/validation` endpoint requires the `execq` socket to work:

https://github.com/wazuh/wazuh/blob/405b3e00b2ddad59e2add28fbf7b2d2988b90db3/framework/wazuh/core/manager.py#L296-L305

This means that endpoints like `PUT /manager/restart` will also fail:

```
2021/04/12 13:47:28 ERROR: Connection to 'execq' socket failed
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/utils.py", line 155, in manager_restart
    conn.connect(socket_path)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/dapi/dapi.py", line 234, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
  File "/var/ossec/framework/python/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/dapi/dapi.py", line 208, in run_local
    data = self.f(**self.f_kwargs)
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/rbac/decorators.py", line 417, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/rbac/decorators.py", line 417, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/manager.py", line 277, in restart
    manager_restart()
  File "/var/ossec/framework/python/lib/python3.8/site-packages/wazuh-4.0.4-py3.8.egg/wazuh/core/cluster/utils.py", line 157, in manager_restart
    raise WazuhInternalError(1902)
wazuh.core.exception.WazuhInternalError: Error 1902 - Connection to 'execq' socket failed
2021/04/12 13:47:28 INFO: wazuh 172.19.0.1 "PUT /manager/restart" done in 55.00000000029104ms: 500

```

Some things to consider:

- This only applies to **active configuration**. The result will not change unless we restart the manager to apply the new configuration.

- This will only fail if **the first active-response block** is disabled. These are the use cases that I tested:

  Two AR blocks are enabled (OK)

  ```xml
  <active-response>
    <disabled>no</disabled>
    <command>host-deny</command>
    <location>defined-agent</location>
    <agent_id>032</agent_id>
    <level>10</level>
    <rules_group>sshd,|pci_dss_11.4,</rules_group>
    <timeout>1</timeout>
  </active-response>

  <active-response>
    <disabled>no</disabled>
    <command>firewall-drop</command>
    <location>local</location>
  </active-response>

  ```

  ```json
  {
    "data": {
      "affected_items": [
        {
          "name": "master-node",
          "status": "OK"
        }
      ],
      "total_affected_items": 1,
      "total_failed_items": 0,
      "failed_items": []
    },
    "message": "Validation was successfully checked in all nodes",
    "error": 0
  }
  ```

  Two AR blocks are disabled (ERROR)

  ```xml
  <active-response>
    <disabled>no</disabled>
    <command>host-deny</command>
    <location>defined-agent</location>
    <agent_id>032</agent_id>
    <level>10</level>
    <rules_group>sshd,|pci_dss_11.4,</rules_group>
    <timeout>1</timeout>
  </active-response>

  <active-response>
    <disabled>no</disabled>
    <command>firewall-drop</command>
    <location>local</location>
  </active-response>

  ```

  ```json
  {
    "title": "Wazuh Internal Error",
    "detail": "Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused",
    "remediation": "Please, restart Wazuh to restore sockets",
    "dapi_errors": {
      "master-node": {
        "error": "Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused",
        "logfile": "WAZUH_HOME/logs/api.log"
      }
    },
    "error": 1013
  }
  ```
  
  The first AR block is disabled (ERROR)

  ```xml
  <active-response>
    <disabled>no</disabled>
    <command>host-deny</command>
    <location>defined-agent</location>
    <agent_id>032</agent_id>
    <level>10</level>
    <rules_group>sshd,|pci_dss_11.4,</rules_group>
    <timeout>1</timeout>
  </active-response>

  <active-response>
    <disabled>no</disabled>
    <command>firewall-drop</command>
    <location>local</location>
  </active-response>

  ```

  ```json
  {
    "title": "Wazuh Internal Error",
    "detail": "Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused",
    "remediation": "Please, restart Wazuh to restore sockets",
    "dapi_errors": {
      "master-node": {
        "error": "Unable to connect with socket: Socket: WAZUH_PATH/queue/alerts/execq. Error Connection refused",
        "logfile": "WAZUH_HOME/logs/api.log"
      }
    },
    "error": 1013
  }
  ```
  
  The second AR block is disabled (OK)

  ```xml
  <active-response>
    <disabled>no</disabled>
    <command>host-deny</command>
    <location>defined-agent</location>
    <agent_id>032</agent_id>
    <level>10</level>
    <rules_group>sshd,|pci_dss_11.4,</rules_group>
    <timeout>1</timeout>
  </active-response>

  <active-response>
    <disabled>no</disabled>
    <command>firewall-drop</command>
    <location>local</location>
  </active-response>

  ```
  
  ```json
  {
    "data": {
      "affected_items": [
        {
          "name": "master-node",
          "status": "OK"
        }
      ],
      "total_affected_items": 1,
      "total_failed_items": 0,
      "failed_items": []
    },
    "message": "Validation was successfully checked in all nodes",
    "error": 0
  }
  ```

This issue is not related to the framework or the API as these endpoints only try to connect to the `execq` socket to send a message. We will need further investigation from the core team.

Regards,
Víctor Fernández

**Comment by antoniomanuelfr** [comment]
Hi all!! 

I have been investigating this and as @vicferpoy mentioned, the problem is caused by how `execd` starts:
- After the configuration is read, `execd` checks if Active Response is disabled. 
- If Active Response is disabled, `execd` exits without creating the `execq` socket.
https://github.com/wazuh/wazuh/blob/23e88df954d9169120bd6d458d003f11f4b37bea/src/os_execd/execd.c#L217-L231

To test this, I have done the following script: 
```python
import socket
execq_socket_path = '/var/ossec/queue/alerts/execq'
execq_socket = socket.socket(socket.AF_UNIX, socket.SOCK_DGRAM)
execq_socket.connect(execq_socket_path)
```
After the script is executed with AR disabled: 
```
root@ubuntumanager:/vagrant# python3 test.py 
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    execq_socket.connect(execq_socket_path)
ConnectionRefusedError: [Errno 111] Connection refused
```

With the same config and calling `StartMQ` before `execd` exits, the socket is created and the script won't fail:
```
root@ubuntumanager:/vagrant# python3 test.py 
```

**Comment by antoniomanuelfr** [comment]
Working branch `7968-move-check-manager-config-to-wcom`

**Comment by davidjiglesias** [comment]
Investigation completed, to be fixed in #8356 and #8360
